File: c/code.c
#include <stdio.h>
#include <cv.h>
#include <highgui.h>

IplImage* image = 0;
IplImage* original_image = 0;
double low = 5, ratio = 3;

void canny(IplImage*,int);
void region_of_interest(IplImage*);
void gauss(IplImage*,int);
void hough(IplImage*,float, float, int, float, float);
void find_lanes(IplImage*, IplImage*);
CvPoint polygon[4];

int main(int argc, char**argv){

	image = cvLoadImage(argv[1], CV_LOAD_IMAGE_GRAYSCALE);
	// original_image = cvLoadImage(argv[1], CV_LOAD_IMAGE_UNCHANGED);
	// gauss(image, 11);
	// canny(image,3);
	// region_of_interest(image);
	// cvShowImage("imagem3", image);
	// hough(image, 1, CV_PI/180, 69, 10, 10);
	// cvShowImage("imagem4", image);
	// find_lanes(image,original_image);
	// cvShowImage("imagem5", image);
	
	// cvShowImage("imagem", original_image);

	CvCapture* video = cvCaptureFromFile("solidWhiteRight.mp4");
	IplImage* original_frame;
	original_frame = cvQueryFrame(video);

	double fps = cvGetCaptureProperty(video, CV_CAP_PROP_FPS);
	CvSize frame_size;
    frame_size.width = original_frame->width;
    frame_size.height = original_frame->height;

	CvVideoWriter* output_video = cvCreateVideoWriter("output_video.mp4", CV_FOURCC('P','I','M','4'), fps, frame_size, 1);

	while(cvGrabFrame(video) > 0){
		printf("dsa ");
		original_frame = cvQueryFrame(video);
		IplImage* frame;

		cvSaveImage("original_frame.jpg", original_frame, 0);
		frame = cvLoadImage("original_frame.jpg", CV_LOAD_IMAGE_GRAYSCALE);

		gauss(frame, 11);
		canny(frame,3);
		region_of_interest(frame);
		hough(frame, 1, CV_PI/180, 69, 10, 10);
		find_lanes(frame,original_frame);

		cvWriteFrame(output_video, original_frame);

		// cvShowImage("frame", original_frame);
	}


	cvWaitKey(0);

	cvReleaseImage(&image);
}

void gauss(IplImage* image, int weight){
	if(weight % 2 == 0)
		printf("\n\nERROR IN: gauss(weight) -> weight should be an ODD number (ex: 1, 3, 5...)\n\n");

	cvSmooth(image, image, CV_GAUSSIAN, weight, weight, 0, 0);
}

void canny(IplImage* image, int apertureSize){
	const CvArr* img = image;
	cvCanny(img, image, low, low*ratio, apertureSize);
}

void region_of_interest(IplImage* image_frame){
	const CvArr* img = image_frame;
	IplImage* mask = cvCreateImage(cvGetSize(image_frame), 8, 1);
    polygon[0] = cvPoint(140,530);
    polygon[1] = cvPoint(430,330);
    polygon[2] = cvPoint(530,330);
    polygon[3] = cvPoint(920,530);
	CvPoint* curveArr[]={polygon};
	int nCurvePts[]={4};
	int nCurves=1;

	cvFillPoly(mask, curveArr, nCurvePts, nCurves, CV_RGB(255,255,255), 8, 0);
	cvAnd(image_frame, mask, image_frame, NULL);

}

void hough(IplImage* image, float rho, float theta, int threshold, float min_line_len, float max_line_gap){
	CvMemStorage* storage = cvCreateMemStorage(0);
	CvSeq* lines = 0;
	lines = cvHoughLines2(image, storage, CV_HOUGH_PROBABILISTIC, rho, theta, threshold, min_line_len, max_line_gap);

	IplImage* image2 = cvCreateImage( cvGetSize(image), 8, 1);
	cvSetZero(image2);

	for( int i = 0; i < lines->total; i++ ) {
        CvPoint* line = (CvPoint*)cvGetSeqElem(lines,i);
        cvLine(image2, line[0], line[1], CV_RGB(255,255,255), 2, 8, 0);
    }

    image = image2;
}

void find_lanes(IplImage* image, IplImage* original_image){
	int upperLimit = polygon[1].y;

	// Finds the first (most to left and top) red pixel on the left side of the image
	int leftLine[4] = {0, 0, 0, 0};
	for (int i = upperLimit; i < image->height; i++){
		if(leftLine[0] != 0)
			break;
		for (int j = 0; j < image->width/2; j++){
			if(CV_IMAGE_ELEM(image,unsigned char,i,j) > 0){
				leftLine[0] = j;
            	leftLine[1] = i;
			}
		}
	}
	
    // Finds the last (most to right and bottom) red pixel on the left side of the image
    for (int i = image->height - 1; i > upperLimit; i--){
    	if(leftLine[2] != 0)
    		break;
		for (int j = 0; j < image->width/2; j++){
			if(CV_IMAGE_ELEM(image,unsigned char,i,j) > 0){
				leftLine[2] = j;
            	leftLine[3] = i;
			}
		}
    }

    // Finds the first (most to left and top) red pixel on the right side of the image
    int rightLine[4] = {0, 0, 0, 0};
	for (int i = upperLimit; i < image->height; i++){
		if(rightLine[0] != 0)
			break;
		for (int j = image->width/2; j < image->width; j++){
			if(CV_IMAGE_ELEM(image,unsigned char,i,j) > 0){
				rightLine[0] = j;
            	rightLine[1] = i;
			}
		}
	}

    // Finds the last (most to right and bottom) red pixel on the right side of the image
    for (int i = image->height - 1; i > upperLimit; i--){
    	if(rightLine[2] != 0)
    		break;
		for (int j = image->width/2; j < image->width; j++){
			if(CV_IMAGE_ELEM(image,unsigned char,i,j) > 0){
				rightLine[2] = j;
            	rightLine[3] = i;
			}
		}
    }

    // Calculates the slopes of each lane line to extrapolate
    float leftSlope = ((float) leftLine[1] - leftLine[3]) / ((float) leftLine[0] - leftLine[2]);
    float left_b = ((float) leftLine[1]) - ((float) (leftSlope * leftLine[0]));
    float left_x = (image->height - left_b) / leftSlope;
    leftLine[2] = (int) left_x;
    leftLine[3] = image->height;

    float rightSlope = ((float) rightLine[1] - rightLine[3]) / ((float) rightLine[0] - rightLine[2]);
    float right_b = ((float) rightLine[1]) - ((float) (rightSlope * rightLine[0]));
    float right_x = (image->height - right_b) / rightSlope;
    rightLine[2] = (int) right_x;
    rightLine[3] = image->height;

    CvPoint leftPointLines[] = {cvPoint(leftLine[0],leftLine[1]), cvPoint(leftLine[2],leftLine[3])};
    CvPoint rightPointLines[] = {cvPoint(rightLine[0],rightLine[1]), cvPoint(rightLine[2],rightLine[3])};

    // CvPoint* lines = {leftPointLines, rightPointLines};

    cvLine(original_image, leftPointLines[0], leftPointLines[1], CV_RGB(255,0,0), 4, CV_AA, 0);
    cvLine(original_image, rightPointLines[0], rightPointLines[1], CV_RGB(255,0,0), 4, CV_AA, 0);

    // return lines;
}

File: python/code.py
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import math
import numpy as np
import cv2
from moviepy.editor import VideoFileClip

lastPoints = None

def process_image(image):
    global lastPoints
    originalImage = np.copy(image)

    # Turns image grayscale
    image = grayscale(image)
    # Applies blur
    image = gaussian_blur(image, 5)
    # Applies canny edge detection
    image = canny(image, 50, 130)
    # Defines the polygon that we are interested in and ignores the rest of the image (image is 960x540)
    image = region_of_interest(image, np.array([[(60,image.shape[0]),(430, 330), (530, 330), (image.shape[1]-60,image.shape[0])]], dtype=np.int32))
    # Uses Hough transform to detect lines in the polygon
    image = hough_lines(image, 1, np.pi / 180, 15, 30, 20)
    # Detects where each lane line starts and ends
    points = np.array(findLaneLinesPoints(image))

    # Smooths line position changes
    if (lastPoints is not None):
        points = (points + 4 * lastPoints) / 5
    lastPoints = points

    # Creates new dark image
    image = image * 0
    # Draws lines over the lane lines
    draw_lines(image, points, [255, 0, 0], 6)
    # Overlays the original image with the image with the lines drawn
    image = weighted_img(image, originalImage, 1., 1., 0.)

    return image


def findLaneLinesPoints(img):
    # This function finds the starting and ending points of each of the lane lines
    # It finds the positions of these specific pixels (marked as an X):
    #        / X       X \
    #       / /         \ \
    #      / /           \ \
    #     / /             \ \
    #    / /               \ \
    #   X /                 \ X

    # Finds the first (most to left and top) red pixel on the left side of the image
    leftLine = [0, 0, 0, 0];
    for j in range(320, img.shape[0]):
        if leftLine[0] != 0:
            break
        for i in range(0, int(img.shape[1] / 2)):
            if img[j, i, 0] > 0:
                leftLine[0] = i
                leftLine[1] = j

    # Finds the last (most to right and bottom) red pixel on the left side of the image
    for j in range(img.shape[0] - 1, 320, -1):
        if leftLine[2] != 0:
            break
        for i in range(0, int(img.shape[1] / 2)):
            if img[j, i, 0] > 0:
                leftLine[2] = i
                leftLine[3] = j
                break

    # Finds the first (most to left and top) red pixel on the right side of the image
    rightLine = [0, 0, 0, 0];
    for j in range(320, img.shape[0]):
        if rightLine[0] != 0:
            break
        for i in range(int(img.shape[1] / 2), img.shape[1]):
            if img[j, i, 0] > 0:
                rightLine[0] = i
                rightLine[1] = j
                break

    # Finds the last (most to right and bottom) red pixel on the right side of the image
    for j in range(img.shape[0] - 1, 320, -1):
        if rightLine[2] != 0:
            break
        for i in range(int(img.shape[1] / 2), img.shape[1]):
            if img[j, i, 0] > 0:
                rightLine[2] = i
                rightLine[3] = j

    # Calculates the slopes of each lane line to extrapolate
    leftSlope = float(leftLine[1] - leftLine[3]) / float(leftLine[0] - leftLine[2])
    b = float(leftLine[1]) - float(leftSlope * leftLine[0])
    x = (img.shape[0] - b) / leftSlope
    leftLine[2] = int(x)
    leftLine[3] = img.shape[0]

    rightSlope = float(rightLine[1] - rightLine[3]) / float(rightLine[0] - rightLine[2])
    b = float(rightLine[1]) - float(rightSlope * rightLine[0])
    x = (img.shape[0] - b) / rightSlope
    rightLine[2] = int(x)
    rightLine[3] = img.shape[0]

    return [[leftLine, rightLine]]


def grayscale(img):
    return cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)


def canny(img, low_threshold, high_threshold):
    return cv2.Canny(img, low_threshold, high_threshold)


def gaussian_blur(img, kernel_size):
    return cv2.GaussianBlur(img, (kernel_size, kernel_size), 0)


def region_of_interest(img, vertices):
    mask = np.zeros_like(img)

    # print(img.shape[0])

    if len(img.shape) > 2:
        channel_count = img.shape[2]
        ignore_mask_color = (255,) * channel_count
    else:
        ignore_mask_color = 255

    cv2.fillPoly(mask, vertices, ignore_mask_color)

    masked_image = cv2.bitwise_and(img, mask)
    return masked_image


def draw_lines(img, lines, color=[255, 0, 0], thickness=4):
    for line in lines:
        for x1, y1, x2, y2 in line:
            cv2.line(img, (x1, y1), (x2, y2), color, thickness)


def hough_lines(img, rho, theta, threshold, min_line_len, max_line_gap):
    lines = cv2.HoughLinesP(img, rho, theta, threshold, np.array([]), min_line_len, max_line_gap)
    line_img = np.zeros((img.shape[0], img.shape[1], 3), dtype=np.uint8)
    draw_lines(line_img, lines)
    return line_img


def weighted_img(img, initial_img, a=0.8, b=1., d=0.):
    return cv2.addWeighted(initial_img, a, img, b, d)

# Opens up image and processes it
# image = mpimg.imread('test_images/solidWhiteRight.jpg')
# print('This image is:', type(image), 'with dimensions:', image.shape)
# image = process_image(image)
# plt.imshow(image)
# plt.show()

# Opens up video and processes it
white_output = '../test_videos_output/solidYellowLeft.mp4'
clip1 = VideoFileClip("../test_videos/solidYellowLeft.mp4")
white_clip = clip1.fl_image(process_image)
white_clip.write_videofile(white_output, audio=False)
